onnx/shape_inference/implementation.cc

```diff
@@ -183,8 +183,34 @@ static void InferShapesImpl(
   }
 
   std::unordered_map<std::string, const TensorProto*> inputDataByName;
+  GraphProto inputGraph; // to not modify original graph; use temp graph  
```

Comment:
I am confused. Why do we need a temp graph? Where is this eventually used?

Reply:
I guess you are using it for automatically allocating and freeing new ValueInfoProto below? I guess we just need a TypeProto, actually. At the least, it is worth documenting.

Reply:
But it might be simpler to just use a vector of TypeProto, or something like that, to do the memory management.

Reply:
> I guess you are using it for automatically allocating and freeing new ValueInfoProto below? I guess we just need a TypeProto, actually. At the least, it is worth documenting.

Yes. I also felt weird about creating a useless graph... Sounds like I can simply use `valueTypesByName[tp.name()]= new TypeProto()` for creating a new TypeProto. Thank you for the good catch.

Reply:
Yes, that would be better ... but then you need to remember to "free" the allocated TypeProto at the end. (That is why I was referring to memory management.) As long as that happens correctly, this would be good.